File: 3614-process-string-with-special-operations-ii/3614-process-string-with-special-operations-ii.py
class Solution:
    def processStr(self, s: str, k: int) -> str:
        size = 0
        for c in s:
            if c == "*":
                if size:
                    size -= 1
            elif c == "#":
                size *= 2
            elif c == "%":
                pass
            else:
                size += 1
        if k + 1 > size:
            return "."
       
        for c in s[::-1]:
            if c == '*':
                size += 1
            elif c == '#':
                if k + 1 > size // 2:
                    k -= size // 2

                size //= 2
            elif c == '%':
                k = size - k - 1
            else:
                if size == k + 1:
                    return c
                
                size -= 1

        return '.'

File: 3614-process-string-with-special-operations-ii/3614-process-string-with-special-operations-ii.java
class Solution {
    public char processStr(String s, long k) {
        long size = 0;

        for (char c : s.toCharArray()){
            if (c == '*'){
                if (size > 0) size -= 1;
            } else if (c == '#'){
                size *= 2;
            } else if (c == '%'){
                continue;
            } else {
                size += 1;
            }
        }

        if (k + 1 > size) return '.';

        for (int i = s.length() - 1; i > -1; i--){
            char c = s.charAt(i);
            if (c == '*'){
                size += 1;
            } else if (c == '#'){
                if (k + 1 > size / 2){
                    k -= size / 2;
                }
                size /= 2;
            } else if (c == '%'){
                k = size - k - 1;
            } else {
                if (size == k + 1) return c;

                size -= 1;
            }
        }

        return '.';
    }
}